hf_nml_outbnd: Minor bug in ww3_grid.ftn: Info about outbound files

CarstenHansen opened the pull request:
In the output from ww3_grid, some info text is missing about the last outbound boundary file for one-way nesting: "File nestI.ww3 Number of points .."  when namelists input files are used.

Solution: Let NPO=0 mark that we have got the info for the last output b.c. file. Then the do-loop over outbound lines will exit immediately after the block that writes the info.

Additional correction: The file ww3_grid.scratch isn't used when namelists input files are used. Therefore, also prepend a condition "IF ( .NOT. FLGNML )" to other statements handling the file NDSS.

CarstenHansen committed:
hf_nml_outbnd: Fix a minor bug in ww3_grid.ftn: Info about outbound files

Implication: Affects the user's ability to check the output of ww3_grid.

The bug exists when namelists input files are used. It consists of two
parts, a and b:

Part a: A code block is not reached that writes info to NDSO about the
last boundary file: "File nestI.ww3 Number of points .." (for I ==
NML_OUTBND_COUNT%N_LINE).

Part b: A code line "WRITE(NDSS,'(A)') NML_OUTBND_LINE(I-1)" writes
garbage to file ww3_grid.scratch, because the NML_OUTBND_LINE elements
are not strings.

Solution part a: Let NPO=0 mark that we have got the info for the last
output b.c. file, and the do-loop over outbound lines will exit
immediately after the block that writes the info.

Solution part b: Ommit the write statement. The file ww3_grid.scratch
isn't used when namelists input files are used. Therefore, also prepend
a condition "IF (.NOT. FLGNML)" to the statements
"OPEN ( NDSS,FILE=TRIM(FNMPRE)//'ww3_grid.scratch',...",
"CLOSE ( NDSS,...", and for REWIND (NDSS) ... NDSI2 = NDSS.

Issue remaining: Update the Manual, sect. "4.4.3 The grid preprocessor"

ajhenrique commented:
Thanks @CarstenHansen. @mickaelaccensi please have a look at the proposed fix in regards to namelists and comment.

mickaelaccensi commented on model/ftn/ww3_grid.ftn:
```diff
@@ -4876,7 +4876,9 @@
 !
       WRITE (NDSO,990)
-      OPEN (NDSS,FILE=TRIM(FNMPRE)//'ww3_grid.scratch',FORM='FORMATTED')
+! ww3_grid.scratch is not used with namelist input (C.Hansen 20190701)
```
to remove and mention in commit message

mickaelaccensi commented on model/ftn/ww3_grid.ftn:
```diff
@@ -4906,21 +4910,29 @@
+! ( The block does WRITE(NDSO,991), WRITE (NDSO,992), etc.)
+! Instead, let NPO=0 mark that we have got the info for the last file, and the
+! do-loop over outbound lines will exit immediately after the block. This is a
+! minor bug because it only affects output messages (C. Hansen 20190701)
```
to remove and mention in the commit message

mickaelaccensi commented on model/ftn/ww3_grid.ftn:
```diff
@@ -4906,21 +4910,29 @@
+!            IF (FLGNML) THEN
+! NML_OUTBND_LINE are not strings (C. Hansen 20190701), but NDSS isn't used
+!              WRITE(NDSS,'(A)') NML_OUTBND_LINE(I-1)
+!            ELSE
```
to remove and mention in the commit message

CarstenHansen replied:
I understand, thank you. I will do as you ask.

mickaelaccensi reviewed (changes requested):
I agree with the bugfix correction with NPO=0 instead of exit statement.
Carsten, can you remove all the your comments and commented lines to clean up the code before the pull request. All explanations of this bugfix must be tracked in the commit message only.

thanks for your help,

ajhenrique commented:
@CarstenHansen can you please weigh in @mickaelaccensi comments above? I would also like to ask that you include the changes proposed by @MicahDavid with your remarks as per PR #57 bundled with your code changes. Thanks!

CarstenHansen commented:
I will include the changes proposed by @MicahDavid and my suggested extension.

mickaelaccensi commented:
bugfix integrated in PR #86 and referenced in issue #89

CarstenHansen committed:
Bugs that affect messages about parsing the input files.

There are three bug fixes (A, B, C) for ww3_grid and ww3_shel in two files ( ww3_grid.ftn and w3iogomd.ftn) .

A) A temporary file ww3_grid.scratch is opened not only for old-stype input file ww3_grid.inp, but also when input in ww3_grid.nml is present. The bugfix does not open this file in the latter case.

B) In the standard output from ww3_grid  info is missing about the last outbound file for one-way nesting. The solution is to let NPO=0 mark that info has been read about the last file, and use this value to exit the file message loop.

C) When ww3_shel is running under MPI and and encounters an error in the list of fields specified in ww3_shel.inp or ww3_shel.nml for the output to out_grd.ww3, it attempts writing a WARNING to a file id NDSE == -1. This causes the program to abort on some systems. @MicahDavid was the first to debug this. A file id NDSE /= -1 is defined only for the parallel process to write output ( IAPROC .EQ. NAPOUT ). The solution is to use a condition "IF (NDSEN .GE. 0 " throughout both the subroutines W3READFLGRD and W3FLGRDFLAG. This condition already exists in the subroutines NEXTLN and W3FLDO. This bugfix will probably also apply for ww3_multi.

CarstenHansen committed:
Add revised ww3_grid.ftn

CarstenHansen commented:
I browsed the file changes in develop #86. It completes this pull request, and also the changes proposed by @MicahDavid and my suggested extension. May I close this pull request?

CarstenHansen committed:
Merge branch 'develop' of https://github.com/NOAA-EMC/WW3 into hf_nml_outbnd

ZZ

ajhenrique commented:
Embedded into Ifremer PR. Closing.
